### PR TITLE
Law Machinery

### LINKED ISSUE #266: Add machinery for laws for effects
I've been thinking about how to prove interpreters are correct. Maybe this:

```haskell
class HasLaws (e :: Effect) where
  laws :: Member e r => [exists a. Eq a => (Gen (Sem r a, Sem r a))]
```

which states that for each element of the list, the two programs should evaluate to the same `a` under any interpreter (obviously using the impredicativty trick for `laws`)

For example:

```haskell
instance Eq s => HasLaws (State s) where
  laws = 
    [ do  -- overwriting of put
        s <- arbitrary
        s' <- arbitrary
        pure (put s >> put s' >> get, put s' >> get)

    , -- more
    ]
```

### PR DESCRIPTION
**Pull request by isovector**
This PR adds machinery for writing easy laws that should hold for an effect. It gives an example of what they'd look like for `State s`.

Fixes #266

### CODE REVIEW

**Comment by googleson78** [comment]
Sorry if I'm missing something (including the point entirely), but can't this live in tests (to avoid depending on `QuickCheck` from the library)?

**Comment by isovector** [comment]
I wish! But `Law` needs to live in the library proper (or some other library) so that any other interpretations of `State` can also prove that they're lawful.

**Comment by ocharles** [comment]
`lens` deals with that by having an external library: https://hackage.haskell.org/package/lens-properties

**Comment by ocharles** [comment]
I guess it's not quite the same though, because each effect has its own laws, so they would be a bit weird being in `polysemy-laws`...

**Comment by isovector** [comment]
Yeah --- I can't think of a good solution here. By putting it directly in the library, at least we can set a good example. Maybe these can be hidden behind a cabal flag to disable them if you really don't want the dependency,

**Comment by spacekitteh** [comment]
Alternative approach: Liquid Haskell

**Comment by spacekitteh** [comment]
Also, this CLEARLY cries out for a Generator effect :P

File: src/Polysemy/Law.hs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    ]
+
+printf :: String -> [String] -> String
+printf str args = splitArgs str
+  where
+    splitArgs :: String -> String
+    splitArgs s =
+      case break (== '%') s of
+        (as, "") -> as
+        (as, drop 1 -> (b : bs))
```
Comment (TheMatten) [inline src/Polysemy/Law.hs]: Just a stylistic question - do we need `ViewPatterns` here?

`drop 1 -> (b : bs)` -> `_ : b : bs`

File: src/Polysemy/Law.hs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  , module Test.QuickCheck
+  ) where
+
+import Control.Arrow (first)
+import Data.Char
+import Polysemy
+import Test.QuickCheck
+
+
+class Citizen r a | r -> a where
```
Comment (TheMatten) [inline src/Polysemy/Law.hs]: We'll probably need to have some comment here, or maybe some different name - personally I had to read through the code to get the idea of what `Citizen` means :slightly_smiling_face:

File: src/Polysemy/Law.hs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+printf :: String -> [String] -> String
+printf str args = splitArgs str
+  where
+    splitArgs :: String -> String
+    splitArgs s =
+      case break (== '%') s of
+        (as, "") -> as
+        (as, drop 1 -> (b : bs))
+          | isDigit b
+          , let d = read [b] - 1
```
Comment (TheMatten) [inline src/Polysemy/Law.hs]: Just a nitpick - this code seems to work with one-digit indexes only - is this a problem?

File: src/Polysemy/State/Law.hs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+
+------------------------------------------------------------------------------
+-- | A collection of laws that show a `State` interpreter is correct.
+prop_lawfulState
+    :: forall r s
+     . (Eq s, Show s, Arbitrary s, MakeLaw (State s) r)
+    => InterpreterFor (State s) r
+    -> Property
+prop_lawfulState i12n = conjoin
```
Comment (TheMatten) [inline src/Polysemy/State/Law.hs]: Nitpick - not sure I understand name `i12n` :smile:

File: src/Polysemy/Law.hs (hunk 3 above)
Comment (isovector) [reply src/Polysemy/Law.hs]: You're right, it does. I think if your law requires more than 9 moving parts, it's probably too complicated :)

File: src/Polysemy/State/Law.hs (hunk 4 above)
Comment (isovector) [reply src/Polysemy/State/Law.hs]: **i** nterpretation **n**, but yeah, fair enough

File: src/Polysemy/Law.hs (hunk 1 above)
Comment (isovector) [reply src/Polysemy/Law.hs]: spoilsport ;)

**Comment by KingoftheHomeless** [comment]
hey! I just started working on a review.

**Comment by isovector** [comment]
sorry! i get too excited. i'll respond to your comments  when you're done

File: src/Polysemy/Law.hs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+      -> String
+      -> res
+      -> Law e r
+
+instance MakeLaw e '[] where
+  mkLaw = Law run
+
+instance MakeLaw e '[Embed IO] where
+  mkLaw = LawIO runM
+
```
[the remark this answers is not in the document]
Comment (isovector) [reply src/Polysemy/Law.hs]: sure! i'm just adding crap here as i need it.

**Comment by isovector** [comment]
@KingoftheHomeless good questions. My current plan is to make it as widely-useful batteries-included as possible. Having more packages just makes my job harder, and doesn't really buy anyone anything on this front.

**Comment by isovector** [comment]
That being said, this isn't release yet, so I could still be talked into moving it if yall are convinced it doesn't belong here.

**Pull request merged**